src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c

```diff
@@ -20,11 +20,119 @@ module_param(kds_echo, int, (S_IRUGO|S_IWUSR));
 MODULE_PARM_DESC(kds_echo,
 		 "enable KDS echo (0 = disable (default), 1 = enable)");
 
+static inline void
+xocl_ctx_to_info(struct drm_xocl_ctx *args, struct kds_ctx_info *info)
+{
+	if (args->cu_index == XOCL_CTX_VIRT_CU_INDEX)
+		info->cu_idx = CU_CTX_VIRT_CU;
+	else
+		info->cu_idx = args->cu_index;
+
+	if (args->flags == XOCL_CTX_EXCLUSIVE)
+		info->flags = CU_CTX_EXCLUSIVE;
+	else
+		info->flags = CU_CTX_SHARED;
+}
+
+static int xocl_add_context(struct xocl_dev *xdev, struct kds_client *client,
+			    struct drm_xocl_ctx *args)
+{
+	struct kds_ctx_info info;
+	xuid_t *uuid;
+	int ret;
+
+	mutex_lock(&client->lock);
+	/* If this client has no opened context, lock bitstream */
+	if (!client->num_ctx) {
+		ret = xocl_icap_lock_bitstream(xdev, &args->xclbin_id);
+		if (ret)
+			goto out;
+		uuid = vzalloc(sizeof(*uuid));
+		if (!uuid) {
+			ret = -ENOMEM;
+			goto out;
+		}
+		uuid_copy(uuid, &args->xclbin_id);
+		client->xclbin_id = uuid;
+	}
+
+	/* Bitstream is locked. No one could load a new one
+	 * until this client close all of the contexts.
+	 */
+	xocl_ctx_to_info(args, &info);
+	ret = kds_add_context(client, &info);
+
+out:
+	if (!client->num_ctx) {
+		vfree(client->xclbin_id);
+		client->xclbin_id = NULL;
+		(void) xocl_icap_unlock_bitstream(xdev, &args->xclbin_id);
+	}
+	mutex_unlock(&client->lock);
+	return ret;
+}
+
+static int xocl_del_context(struct xocl_dev *xdev, struct kds_client *client,
+			    struct drm_xocl_ctx *args)
+{
+	struct kds_ctx_info info;
+	xuid_t *uuid;
+	int ret = 0;
+
+	mutex_lock(&client->lock);
+
+	uuid = client->xclbin_id;
+	/* xclCloseContext() would send xclbin_id and cu_idx.
+	 * Be more cautious while delete. Do sanity check */
+	if (!uuid) {
+		userpf_err(xdev, "No context was opened");
+		ret = -EINVAL;
+		goto out;
+	}
+
+	/* If xclbin id looks good, unlock bitstream should not fail. */
+	if (!uuid_equal(uuid, &args->xclbin_id)) {
+		userpf_err(xdev, "Try to delete CTX on wrong xclbin");
+		ret = -EBUSY;
+		goto out;
+	}
+
+	xocl_ctx_to_info(args, &info);
+	ret = kds_del_context(client, &info);
+	if (ret)
+		goto out;
+
+	/* unlock bitstream if there is no opening context */
+	if (!client->num_ctx) {
+		vfree(client->xclbin_id);
+		client->xclbin_id = NULL;
+		(void) xocl_icap_unlock_bitstream(xdev, &args->xclbin_id);
+	}
+
+out:
+	mutex_unlock(&client->lock);
+	return ret;
+}
+
 static int xocl_context_ioctl(struct xocl_dev *xdev, void *data,
 			      struct drm_file *filp)
 {
-	/* TODO: implement this */
-	return 0;
+	struct drm_xocl_ctx *args = data;
+	struct kds_client *client = filp->driver_priv;
+	int ret = 0;
+
+	switch(args->op) {
+	case XOCL_CTX_OP_ALLOC_CTX:
+		ret = xocl_add_context(xdev, client, args);
+		break;
+	case XOCL_CTX_OP_FREE_CTX:
+		ret = xocl_del_context(xdev, client, args);
+		break;
+	default:
+		ret = -EINVAL;
+	}
```

Comment:
It is legal. But it is just my suggestion we add a break for every case including default. :) The default case could be not the last case, though rare.

Reply:
fixed